src/mongocxx/test/spec/monitoring.cpp

```diff
@@ -28,63 +28,39 @@ MONGOCXX_INLINE_NAMESPACE_BEGIN
 namespace spec {
 
 using namespace mongocxx;
+using bsoncxx::to_json;
 
 void apm_checker::compare(bsoncxx::array::view expectations,
                           bool allow_extra,
                           const test_util::match_visitor& match_visitor) {
-    using bsoncxx::to_json;
+    auto is_kill_cursor = [](bsoncxx::document::value v) {
+        return v.view()["command_started_event"]["command"]["killCursors"];
+    };
 
     auto events_iter = _events.begin();
+    auto events_end = (_skip_kill_cursors)
+                          ? std::remove_if(_events.begin(), _events.end(), is_kill_cursor)
+                          : _events.end();
+
     for (auto expectation : expectations) {
-        if (events_iter == _events.end()) {
-            fprintf(stderr,
-                    "more expectations than events captured\nExpectations:\n\n%s\n\nCaptured:\n\n",
-                    to_json(expectations).c_str());
-            this->print_all();
-            REQUIRE(false);
-        }
-        REQUIRE(events_iter != _events.end());
+        REQUIRE(events_iter != events_end);
 
         auto expected = expectation.get_document().view();
+        CAPTURE(to_json(*events_iter), expectation);
         REQUIRE_BSON_MATCHES_V(*events_iter, expected, match_visitor);
         events_iter++;
     }
-    if (!allow_extra) {
-        if (_skip_kill_cursors && events_iter != _events.end()) {
-            if (events_iter->view()["command_started_event"]["command"]["killCursors"]) {
-                ++events_iter;
-            }
-        }
-
-        if (events_iter == _events.end()) {
-            return;
-        }
-
-        if (events_iter != _events.end()) {
-            fprintf(stderr, "Error: extra event captured: %s\n", to_json(*events_iter).c_str());
-            events_iter++;
-        }
-
-        REQUIRE(false);
-    }
+
+    REQUIRE((allow_extra || events_iter == events_end));
```

Comment:
We no longer have to worry about kill cursors here. So we either allow extra events or events_iter is at the "end" (which may be the start of kill cursor commands if we skipped them). If this fails we know we have too many events and the line number will be printed.

Reply:
As a sidenote, with the removal of kill cursors events, we might be able to remove `allow_extra`. Though other spec tests set it too, so I'm not sure off-hand.

Reply:
Just from trying it locally, it causes tests to break though I'm not sure why. I'll investigate

Reply:
Thanks for checking. There's no need to investigate further IMO. It's out of scope for this change.